core/platform/image/gae_image_services_test.py

```diff
@@ -0,0 +1,105 @@
+# coding: utf-8
+#
+# Copyright 2018 The Oppia Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Tests for methods in the gae_image_services."""
+
+from __future__ import absolute_import  # pylint: disable=import-only-modules
+from __future__ import unicode_literals  # pylint: disable=import-only-modules
+
+import os
+
+from constants import constants
+from core.platform.image import gae_image_services
+from core.tests import test_utils
+import feconf
+import python_utils
+
+
+class GaeImageServicesUnitTests(test_utils.GenericTestBase):
+    """Tests for gae_image_services."""
+
+    TEST_IMAGE_WIDTH = 3000
+    TEST_IMAGE_HEIGHT = 2092
+
+    def setUp(self):
+        super(GaeImageServicesUnitTests, self).setUp()
+        with python_utils.open_file(
+            os.path.join(feconf.TESTS_DATA_DIR, 'dummyLargeImage.jpg'),
+            mode='rb', encoding=None) as f:
+            self.raw_image = f.read()
+
+    def test_get_image_dimensions(self):
+        compressed_image = (
+            gae_image_services.compress_image(self.raw_image, 0.5))
+        height, width = (
+            gae_image_services.get_image_dimensions(compressed_image))
+        # The google app engine images api sometimes rounds off by 1 pixel
+        # compared to the pillow implementation.
+        self.assertTrue(abs(self.TEST_IMAGE_HEIGHT - height) <= 1)
+        self.assertTrue(abs(self.TEST_IMAGE_WIDTH - width) <= 1)
+
+    def test_dev_mode(self):
+        height, width = gae_image_services.get_image_dimensions(self.raw_image)
+        # The google app engine images api sometimes rounds off by 1 pixel
+        # compared to the pillow implementation.
+        self.assertTrue(abs(self.TEST_IMAGE_HEIGHT - height) <= 1)
+        self.assertTrue(abs(self.TEST_IMAGE_WIDTH - width) <= 1)
+
+    def test_compress_image(self):
+        with self.swap(constants, 'DEV_MODE', False):
+            compressed_image = (
+                gae_image_services.compress_image(self.raw_image, 0.5))
+        height, width = (
+            gae_image_services.get_image_dimensions(compressed_image))
+        # The google app engine images api sometimes rounds off by 1 pixel
+        # compared to the pillow implementation.
+        self.assertTrue(abs(int(self.TEST_IMAGE_HEIGHT * 0.5) - height) <= 1)
+        self.assertTrue(abs(int(self.TEST_IMAGE_WIDTH * 0.5) - width) <= 1)
+
+    def test_large_decompress_image(self):
```

Comment:
Just a check -- it's great that you noticed that compress_image() can actually do the reverse (i.e. enlarging)! 

But, if all the callsites pass in a scale factor <= 1, then another acceptable solution is to have the compress_image() function raise an error if the scale factor is > 1. That would lead to less code to maintain and simplification of logic. Just an alternative to consider if the callsites don't require enlarging.

Reply:
So I checked the callsites and there is no calls for enlarging the image but since the original api allowed it, I just added a separate function to only allow enlarging.

Reply:
Hmm. That's fair, but actually I would recommend that you change the API to only allow downsizing. It is less code to maintain and you've already done the due diligence to verify that enlarging is not required. (It hasn't been needed the past several years so I doubt it'll be needed in the future.)

Reply:
Also enlarging tends to result in images that look bad anyway, since you don't have enough pixel density.

Reply:
Ok sounds good I'll just remove it then.